server/src/main/java/com/cloud/resource/ResourceManagerImpl.java

```diff
@@ -3266,12 +3263,16 @@ public List<HostVO> listAllHostsInAllZonesByType(final Type type) {
 
     @Override
     public List<HypervisorType> listAvailHypervisorInZone(final Long zoneId) {
-        List<VMTemplateVO> systemVMTemplates = _templateDao.listAllReadySystemVMTemplates(zoneId);
-        final Set<HypervisorType> hypervisors = new HashSet<>();
-        for (final VMTemplateVO systemVMTemplate : systemVMTemplates) {
-            hypervisors.add(systemVMTemplate.getHypervisorType());
+        final SearchCriteria<String> sc = _hypervisorsInDC.create();
+        if (zoneId != null) {
+            sc.setParameters("dataCenter", zoneId);
         }
-        return new ArrayList<>(hypervisors);
+        sc.setParameters("type", Host.Type.Routing);
+
+        return _hostDao.customSearch(sc, null).stream()
```

Comment:
can list only unique hypervisor entries here?

Reply:
@sureshanaparti SearchBuilder used for this already has DISTINCT statement
https://github.com/apache/cloudstack/blob/4.20/server/src/main/java/com/cloud/resource/ResourceManagerImpl.java#L2034

Reply:
ok, thanks for checking.